mongosql/src/algebrizer/definitions.rs

```diff
@@ -475,14 +477,61 @@ impl<'a> Algebrizer<'a> {
         }
 
         // Build the Project Stage using the source and built expression.
-        let stage = mir::Stage::Project(mir::Project {
+        let project_stage = mir::Stage::Project(mir::Project {
             source: Box::new(source),
-            expression,
+            expression: expression.clone(),
             is_add_fields,
             cache: SchemaCache::new(),
         });
-        stage.schema(&self.schema_inference_state())?;
-        Ok(stage)
+        project_stage.schema(&self.schema_inference_state())?;
+        Ok(if is_distinct {
+            let group_stage = Self::build_distinct_values_group_stage(
+                &expression,
+                project_stage,
+                self.scope_level,
+            );
+            group_stage.schema(&self.schema_inference_state())?;
+            group_stage
+        } else {
+            project_stage
+        })
+    }
+
+    fn build_distinct_values_group_stage(
+        project_tuple: &BindingTuple<Expression>,
+        source_stage: mir::Stage,
+        scope_level: u16,
+    ) -> mir::Stage {
+        let mut group_keys: Vec<OptionallyAliasedExpr> = Vec::new();
+
+        for (key, expr) in project_tuple.iter() {
+            if key.datasource == DatasourceName::Bottom && key.scope == scope_level {
```

Comment:
I think checking the scope is correct -- we can try some unit or e2e tests that use `SELECT DISTINCT` in subquery contexts to test that -- but I'm not convinced we should only be considering the `Bottom` datasource.

Consider `SELECT DISTINCT VALUES foo.*, bar.* FROM foo JOIN bar`. There is no `Bottom` datasource in that query. Instead, there are 2 `Named` datasources that we should use as the group keys.

Reply:
This was super helpful.  I ran the query and it didn't work the way it was implemented.  I changed the logic to use the datasources for the group keys.  I also added a test that has a nested subquery.